### PR TITLE
[Lang] Support to destroy a SnodeTree and manage the memory

### PR DESCRIPTION
**Pull request by beginlner**
Related issue = #2506
This PR support to destroy a `SNodeTree` by `x.destroy()` after `x=<FieldsBuilder>.finalize()`.
Example:
```
fb = ti.FieldsBuilder()
a = ti.field(ti.f64)
fb.dense(ti.i, 100).place(a)
x = fb.finalize()
x.destroy()
```
PS: It's available only with LLVM backend.
Details:
We use a `SNodeTreeBufferManager` class to manage the memory of destroyed `SNodeTree`.
In `SNodeTreeBufferManager`, we keep a `std::set` to buffer memory of destroyed `SNodeTree` in order of size, and a `std::map` in order of start address.

### CODE REVIEW

File: taichi/struct/snode_tree.h [hunk 1]
```diff
@@ -15,30 +17,46 @@ namespace lang {
...
    * @param id Id of the tree
    * @param root Root of the tree
    */
-  explicit SNodeTree(int id, std::unique_ptr<SNode> root, bool packed);
+  explicit SNodeTree(int id,
+                     std::unique_ptr<SNode> root,
+                     bool packed,
+                     Program *prog);
+
+  void destroy();
```
Comment (k-ye) [inline taichi/struct/snode_tree.h:33]: Could you add this as a method in `Program` itself? When destroying a `SNodeTree`, we can just free its memory and remove the tree from the program.

Comment (beginlner) [reply taichi/struct/snode_tree.h:33]: `Destoryed` flag is not necessary. We can maintain it in `SNodeTreeBufferManager`. But `c=fb.finalize()` returns a pointer to the `SNodeTree` and `c.destroy()' destroys it. Then `destroy` method is needed in `SNodeTree`.

Comment (k-ye) [reply taichi/struct/snode_tree.h:33]: You could have the python`SNodeTree` invoke the CPP `Program::destroy_snode_tree()` :-)

Comment (beginlner) [reply taichi/struct/snode_tree.h:33]: Oh, make sense.

File: python/taichi/lang/snode_tree.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        if self.destroyed:
+            raise InvalidOperationError('SNode tree has been destroyed')
+        _ti_core.destroy_snode_tree(self.ptr.id(), impl.get_runtime().prog)
+        self.destroyed = True
+
+    @property
+    def id(self):
+        if self.destroyed:
+            raise InvalidOperationError('SNode tree has been destroyed')
+        return self.ptr.id()
```
Comment (ljcc0930) [inline python/taichi/lang/snode_tree.py:26]: move this to `python/taichi/snode`?

File: taichi/python/export_lang.cpp [hunk 3]
```diff
@@ -194,7 +194,10 @@ void export_lang(py::module &m) {
       .def_readwrite("quant_opt_store_fusion",
                      &CompileConfig::quant_opt_store_fusion)
       .def_readwrite("quant_opt_atomic_demotion",
-                     &CompileConfig::quant_opt_atomic_demotion);
+                     &CompileConfig::quant_opt_atomic_demotion)
+      /*.def_readwrite("memory_allocate_critical_size",
+                     &CompileConfig::memory_allocate_critical_size)*/
```
Comment (ljcc0930) [inline taichi/python/export_lang.cpp:199]:
```suggestion
```

File: taichi/python/export_lang.cpp [hunk 4]
```diff
@@ -866,10 +871,16 @@ void export_lang(py::module &m) {
...
-        });
+  m.def(
+      "finalize_snode_tree",
+      [](SNodeRegistry *registry, const SNode *root,
+         Program *program) -> SNodeTree * {
+        return program->add_snode_tree(registry->finalize(root));
+      },
+      py::return_value_policy::reference);
+  m.def("destroy_snode_tree", [](const int id, Program *program) {
+    program->destroy_snode_tree(id);
```
Comment (ljcc0930) [inline taichi/python/export_lang.cpp:882]:
```suggestion
  m.def("destroy_snode_tree", [](SNodeTree *ptr, Program *program) {
    program->destroy_snode_tree(id);
```
Pass in a `SNodeTree` pointer?

File: taichi/runtime/llvm/runtime.cpp [hunk 5]
```diff
@@ -894,13 +901,17 @@ void runtime_initialize_snodes(LLVMRuntime *runtime,
...
+                               std::size_t rounded_size,
+                               Ptr ptr) {
   // For Metal runtime, we have to make sure that both the beginning address
   // and the size of the root buffer memory are aligned to page size.
-  runtime->root_mem_sizes[snode_tree_id] =
+  /*runtime->root_mem_sizes[snode_tree_id] =
       taichi::iroundup((size_t)root_size, taichi_page_size);
   runtime->roots[snode_tree_id] = runtime->allocate_aligned(
-      runtime->root_mem_sizes[snode_tree_id], taichi_page_size);
+      runtime->root_mem_sizes[snode_tree_id], taichi_page_size);*/
```
Comment (ljcc0930) [inline taichi/runtime/llvm/runtime.cpp:912]: Delete this?

File: taichi/system/snode_tree_buffer_manager.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+               const int snode_tree_id);
+
+  void destroy(const int snode_tree_id);
+
+ private:
+  std::set<std::pair<std::size_t, Ptr>> size_set;
+  std::map<Ptr, std::size_t> Ptr_map;
+  Program *prog;
+  Ptr roots[taichi_max_num_snode_trees];
+  std::size_t sizes[taichi_max_num_snode_trees];
```
Comment (ljcc0930) [inline taichi/system/snode_tree_buffer_manager.h:33]:
```suggestion
  std::set<std::pair<std::size_t, Ptr>> _size_set;
  std::map<Ptr, std::size_t> _Ptr_map;
  Program *_prog;
  Ptr _roots[taichi_max_num_snode_trees];
  std::size_t _sizes[taichi_max_num_snode_trees];
```

File: taichi/system/snode_tree_buffer_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  while (Ptr_map.lower_bound(ptr) != Ptr_map.begin()) {
+    std::pair<Ptr, std::size_t> x = *--Ptr_map.lower_bound(ptr);
+    if (x.first + x.second != ptr) {
+      break;
+    }
+    size_set.erase(std::make_pair(x.second, x.first));
+    Ptr_map.erase(x.first);
+    ptr = x.first;
+    size += x.second;
+  }
```
Comment (ljcc0930) [inline taichi/system/snode_tree_buffer_manager.cpp:28]: Each of the left and right merges would happen at most once?

File: taichi/system/snode_tree_buffer_manager.cpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  Ptr_map[ptr] = size;
+}
+
+Ptr SNodeTreeBufferManager::allocate(JITModule *runtime_jit,
+                                     void *runtime,
+                                     std::size_t size,
+                                     std::size_t alignment,
+                                     const int snode_tree_id) {
+  TI_TRACE("allocating memory for SNode Tree {}", snode_tree_id);
+  if (size_set.lower_bound(std::make_pair(size, nullptr)) == size_set.end()) {
```
Comment (ljcc0930) [inline taichi/system/snode_tree_buffer_manager.cpp:39]:
```suggestion
  auto set_ptr = size_set.lower_bound(std::make_pair(size, nullptr));
  if (set_ptr == size_set.end()) {
```

File: taichi/system/snode_tree_buffer_manager.cpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    runtime_jit->call<void *, std::size_t, std::size_t>(
+        "runtime_snode_tree_allocate_aligned", runtime, size, alignment);
+    auto ptr = prog->fetch_result<Ptr>(taichi_result_buffer_runtime_query_id);
+    roots[snode_tree_id] = ptr;
+    sizes[snode_tree_id] = size;
+    return ptr;
+  } else {
+    std::pair<std::size_t, Ptr> x =
+        *size_set.lower_bound(std::make_pair(size, nullptr));
```
Comment (ljcc0930) [inline taichi/system/snode_tree_buffer_manager.cpp:48]:
```suggestion
    auto x = *set_ptr;
```

File: taichi/program/program.cpp [hunk 10]
```diff
@@ -410,9 +412,17 @@ void Program::initialize_llvm_runtime_snodes(const SNodeTree *tree,
   const int root_id = tree->root()->id;
 
   TI_TRACE("Allocating data structure of size {} bytes", scomp->root_size);
-  runtime_jit->call<void *, std::size_t, int, int>(
+  /*runtime_jit->call<void *, std::size_t, int, int, int>(
       "runtime_initialize_snodes", llvm_runtime, scomp->root_size, root_id,
-      (int)snodes.size(), tree->id());
+      (int)snodes.size(), tree->id());*/
```
Comment (ljcc0930) [inline taichi/program/program.cpp:417]:
```suggestion
```

Comment (beginlner) [reply taichi/program/program.cpp]: All these bugs are fixed.

**Comment by ljcc0930** [comment]
LGTM!

**Review by ljcc0930 — APPROVED** [review]

**Pull request merged**